### ISSUE TITLE
Checking if parameter is not null before setting it up in HashTable for Exchange and other resources)

### ISSUE THREAD

**Issue by PrzemyslawKlys** [microsoft/Microsoft365DSC#653]
My organization is trying to utilize Microsoft365DSC for a large number of Office 365 tenants. As part of this when playing with ExchangeOnline we found out that we don't want some parameters to be set - for example, TargetOwaUrl. If you leave that parameter empty or set it to null things are not accepted and the code throws errors. 

The fix to this is verification if a parameter has value and if it does set it, and if it doesn't skip it.Here's before/after:

![image](https://user-images.githubusercontent.com/15063294/85155562-3046c800-b259-11ea-87d9-b958d9fbf615.png)

Or my preferred way:

```powershell
    foreach ($Key in [string[]]$SetOrganizationRelationshipParams.Keys) {
        if ($null -eq $SetOrganizationRelationshipParams[$Key] -or $SetOrganizationRelationshipParams[$Key] -eq '') {
            $SetOrganizationRelationshipParam.Remove("$Key")
        }
    }
```

Basically removing any key that has a value of null. This makes sure that final commands such as: `New-OrganizationRelationship @NewOrganizationRelationshipParams` or `Set-OrganizationRelationship @SetOrganizationRelationshipParams` don't throw errors when you submit parameters that we don't want set.

So before I go and submit PR to fix some of the resources we're using I wanted to discuss if it makes any sense or maybe there's an alternative to what we're trying to achieve here.

**Comment by NikCharlebois** [comment]
Adding our Exchange lead on the thread @ahohenschuh even if this seems to be larger than just EXO. 

@PrzemyslawKlys What scenario is failing for you exactly? When doing an export, even if the Get-TargetResource function returns all parameters, even those with a $null value, the Export-TargetResource function will remove them from the extracted .ps1 and therefore they should not show. Can you provide more details as to what the ask is here please? Thanks

**Comment by NikCharlebois** [comment]
The error you describe above would only happen if you had manually written the configuration and added $null values manually it appears

**Comment by PrzemyslawKlys** [comment]
Whatever you do with Export-TargetResource or Get-TargetResource doesn't matter. You're using splatting. And splatting if you define parameter of any kind it will get passed as null.

Consider this:

```powershell
function Test-me {
    [cmdletBinding()]
    param(
        [ValidateNotNull()]$Test,
        [string] $Test1
    )
    $Ops
}

$Test5 = 'oops'

Test-me -Test1 $Test5


$Splat = @{
    Test = $NotExistingParameter
    Test1 = $Test5
}

Test-me @Splat
```

The 1st one will not error. 2nd one will because you told Splat to set Test. This is why one of my colleagues (who's code I pasted) check each parameter for null/0 and so on and if it's not null set it. My other solution is to define splat as you did and simply check if each element of splat is empty or null and if it is - remove it. Saves on time and effort on doing 50 If/else checks.

```
    foreach ($Key in [string[]]$SetOrganizationRelationshipParams.Keys) {
        if ($null -eq $SetOrganizationRelationshipParams[$Key] -or $SetOrganizationRelationshipParams[$Key] -eq '') {
            $SetOrganizationRelationshipParam.Remove("$Key")
        }
    }
```

You have to fix it everywhere if you use Splatting and for any reason, parameter is not mandatory. Once you define a splat, you basically pass nulls on parameters that are not defined. 

Hope it clarifies?

So for 50 parameters in those splats

https://github.com/PrzemyslawKlys/Microsoft365DSC/blob/41d1be245c55e1364480fb81929f3f45fa7555b5/Modules/Microsoft365DSC/DSCResources/MSFT_EXOOwaMailboxPolicy/MSFT_EXOOwaMailboxPolicy.psm1#L346-L509

you either need to check if each field is set if it is set it, and if it doesn't skip it. Or you have to use my other suggested approach. 

Also, as I have your attention would it make sense to define two separate splats if they only differ with variable name? 

```
    $NewOrganizationRelationshipParams = @{
        Confirm               = $false
    }
    if ($ArchiveAccessEnabled) {$NeworganizationRelationshipParams.ArchiveAccessEnabled = $ArchiveAccessEnabled}
    if ($DeliveryReportEnabled) {$NeworganizationRelationshipParams.DeliveryReportEnabled = $DeliveryReportEnabled}
    if ($DomainNames) {$NeworganizationRelationshipParams.DomainNames = $DomainNames}
    if ($Enabled) {$NeworganizationRelationshipParams.Enabled = $Enabled}
    if ($FreeBusyAccessEnabled) {$NeworganizationRelationshipParams.FreeBusyAccessEnabled = $FreeBusyAccessEnabled}   
    if ($FreeBusyAccessLevel) {$NeworganizationRelationshipParams.FreeBusyAccessLevel = $FreeBusyAccessLevel}
    if ($FreeBusyAccessScope) {$NeworganizationRelationshipParams.FreeBusyAccessScope = $FreeBusyAccessScope}
    if ($MailboxMoveEnabled) {$NeworganizationRelationshipParams.MailboxMoveEnabled = $MailboxMoveEnabled}
    if ($MailTipsAccessEnabled) {$NeworganizationRelationshipParams.MailTipsAccessEnabled = $MailTipsAccessEnabled}
    if ($MailTipsAccessLevel) {$NeworganizationRelationshipParams.MailTipsAccessLevel = $MailTipsAccessLevel}
    if ($MailTipsAccessScope) {$NeworganizationRelationshipParams.MailTipsAccessScope = $MailTipsAccessScope}
    if ($Name) {$NeworganizationRelationshipParams.Name = $Name}
    if ($OrganizationContact) {$NeworganizationRelationshipParams.OrganizationContact = $OrganizationContact}
    if ($PhotosEnabled) {$NeworganizationRelationshipParams.PhotosEnabled = $PhotosEnabled}
    if ($TargetApplicationUri) {$NeworganizationRelationshipParams.TargetApplicationUri = $TargetApplicationUri}
    if ($TargetAutodiscoverEpr) {$NeworganizationRelationshipParams.TargetAutodiscoverEpr = $TargetAutodiscoverEpr}
    if ($TargetOwaURL) {$NeworganizationRelationshipParams.TargetOwaURL = $TargetOwaURL}
    if ($TargetSharingEpr) {$NeworganizationRelationshipParams.TargetSharingEpr = $TargetSharingEpr}

    $SetOrganizationRelationshipParams = @{
        Confirm               = $false
    }
    if ($ArchiveAccessEnabled) {$SetorganizationRelationshipParams.ArchiveAccessEnabled = $ArchiveAccessEnabled}
    if ($DeliveryReportEnabled) {$SetorganizationRelationshipParams.DeliveryReportEnabled = $DeliveryReportEnabled}
    if ($DomainNames) {$SetorganizationRelationshipParams.DomainNames = $DomainNames}
    if ($Enabled) {$SetorganizationRelationshipParams.Enabled = $Enabled}
    if ($FreeBusyAccessEnabled) {$SetorganizationRelationshipParams.FreeBusyAccessEnabled = $FreeBusyAccessEnabled}    
    if ($FreeBusyAccessLevel) {$SetorganizationRelationshipParams.FreeBusyAccessLevel = $FreeBusyAccessLevel}
    if ($FreeBusyAccessScope) {$SetorganizationRelationshipParams.FreeBusyAccessScope = $FreeBusyAccessScope}
    if ($MailboxMoveEnabled) {$SetorganizationRelationshipParams.MailboxMoveEnabled = $MailboxMoveEnabled}
    if ($MailTipsAccessEnabled) {$SetorganizationRelationshipParams.MailTipsAccessEnabled = $MailTipsAccessEnabled}
    if ($MailTipsAccessLevel) {$SetorganizationRelationshipParams.MailTipsAccessLevel = $MailTipsAccessLevel}
    if ($MailTipsAccessScope) {$SetorganizationRelationshipParams.MailTipsAccessScope = $MailTipsAccessScope}
    if ($Name) {$SetorganizationRelationshipParams.Identity = $Name}
    if ($OrganizationContact) {$SetorganizationRelationshipParams.OrganizationContact = $OrganizationContact}
    if ($PhotosEnabled) {$SetorganizationRelationshipParams.PhotosEnabled = $PhotosEnabled}
    if ($TargetApplicationUri) {$SetorganizationRelationshipParams.TargetApplicationUri = $TargetApplicationUri}
    if ($TargetAutodiscoverEpr) {$SetorganizationRelationshipParams.TargetAutodiscoverEpr = $TargetAutodiscoverEpr}
    if ($TargetOwaURL) {$SetorganizationRelationshipParams.TargetOwaURL = $TargetOwaURL}
    if ($TargetSharingEpr) {$SetorganizationRelationshipParams.TargetSharingEpr = $TargetSharingEpr}
```

Would you accept PRs fixing that?

**Comment by PrzemyslawKlys** [comment]
I'll add that above code if ($Something) should be fixed anyways because it would ignore if something would be set to $false or 0 as well and that's not what want. We want to just prevent $null or empty from being passed, but still $false or 0 are ok.

**Comment by PrzemyslawKlys** [comment]
@NikCharlebois so what do you think?

**Comment by NikCharlebois** [comment]
Sorry for the delay. I see what you mean now, this is specific to the Set-TargetResource. I agree with the dynamic approach where we have the foreach loop. If you do submit a PR, I'd recommend wrapping this loop into a function inside the M365DSCUtil.psm1 module instead of having the code repeated for every resource. Great catch Przemyslaw
